### ISSUE TITLE
General code style issues

### ISSUE THREAD

**Issue by mknjc** [philbowles/asyncmqtt#1]
Hi philbowles,

this project looks very promising but I found a few bad things shortly looking at your code:

- https://github.com/philbowles/asyncmqtt/blob/8d3303a5d40304ba47597b07ce62b306514015ac/src/AsyncMQTT.h#L9 please never "use namespace" in any header a user of your library is supposed to include. This might break a lot of things.
- https://github.com/philbowles/asyncmqtt/blob/8d3303a5d40304ba47597b07ce62b306514015ac/src/AsyncMQTT.h#L27 are macros like this really needed? I can't see any benefit of it but reading it is quite confusing
- Your library needs the Arduino framework (for IPAddress) but uses std::string instead of Arduino Strings. This loads a lot of conversion overhead onto the library users. Is there any reason for it?
- Why are fields like "_username" static and fields like "_port" not? This prevents having multiple mqtt connections.  
- In a few functions your code looses a lot clarity for example in AsyncMQTT::_onData "p" is a really bad name, "bits" should better named something like "headerFlags", "grl" for your remainingLengh + remainingLengthLength pair is really bad and a lot magic bit shifts could better be done with named constants.
- Your complete packet parsing code just assumes the packet is fully in the buffer. But neither AsyncEspTCP nor TCP it self guarantees you this. This could easy crash your code.
- Additionally you assume packet lengths without checking if your buffer could contain the data or the packet size (remaining length) is correct. This allows easy ways for an attacker to crash your code.
- I think your "garbage" Packet ownership code is very complicated and not really the best design for something like you try to achieve there. Additionally writing "new XYZ()" without a variable holding the pointer looks very much like a memory leak through you unorthodox design.

Keep up the good work, I hope your library will evolve in a very good replacement for marvinroger/async-mqtt-client

**Comment by bertmelis** [comment]
Additionally: I'd make the inheritance from AsyncClient private.

Looks promising!

**Comment by philbowles** [comment]
@mknjc @bertmelis  Thanks both for the comments. Some of them regarding style are accepted: I knocked this up in four days, so corners were cut.

"Style" is mostly a personal thing. I don't care if code *looks* like crap as long as it *works*. If it works and looks nice, it's a bonus but if it works, it works. End of.

However, I do take your point(s) and will consider them now I have a working version. Waht I REALLY appreciate though is mknjc's comment re "looking like a memory leak" and I'm going to discuss in conjunction with other memory management being "too complex".

I actually had a comment in there originally saying "This looks like a memory but it is NOT!" to remind *myself*, such is the degreee to which In well understand how it *looks*! So your comments made me actually laugh out loud, in a good way. It is PRECISELY this "oddness" that a) allows the code to work and b) is the root cause of why "the other library" doesn't (and can't and never will) Exactly because it *doesn't* come out of "C101 for Dummies" like all the rest of his code does - because simple code doesn't always work on complex problems. *Especially* when async and timing rear their ugly heads. I'd suggest that unless you know the MQTT spec inside out (I do *now*, I didn't last week !) you may wonder why some of the things are done the way they are. It is quite "tricky".

I started writing bare-metal real-time async code on commercial mainframes in 1975 *without an OS*, so I know a thing or two about it :) The point I'm making here is that I usually *do* know what I'm doing  even if the the code at first sight might not "look" like "professional" code. If I *wanted* it to, it would: I know enough on *how* to do that if I had the time - or cared. 

The pointer (sorry, mknj, you're gonna lose on this one, "p" is an *excellent* name for a short-lived pointer) which is passed to AsyncTCP *must* stay alive until either it is ACKed by MQTT or b) some keepalive timeout has passed, in which case, it has to be safely discarded. Given that I cannot know ahead of time what I need or when - all inside of a resource limited MCU, I have to allocate it dynamically and I have to have total control over when it is freed.

As you have seen I have given "all power to the packet". The packets builds itself according to its type, send itself and puts itself into the correct inbound or outbound queue. It also knows what it took to construct it, therefore it alone knows what to free / undo when it is told "time is up". So, the packet has its own internal memory manager, Yes it is complex, but it is specifically designed *not* to look like any "standard" technique, because "standard" techniques don't work here or we wouldn't be having this conversation.

Also, you should already know just by looking @ 622 lines that work and do MORE than  the 1800 line pile of shit they replace is certainly not got to be doing anything it doesnt *need* to. *Every* single one of those 622 is there for a specific and carefully thought out reason and "earns its keep" *and* can be explained and justified (even if you dont necessarily agree)

Then there is the internal garbage collector for tiny packets that don't need a complex build or deliberately "hanging" data: these are sent by copy not reference so there is nothing to save so they have their own faster, more memory efficiecnt system, to keep the heap size amd garbage collection overhead down during busy peaks. Ya know, the kind of "little things" that many other library authros don't even know they need, let alone have code to handle...

Next there are TWO queues, inbound & outbound. *NO* code that has only 1 queue can  correctly hande Qos2 becaue the server has its own message ids that can be (and usually always *ARE*) the same as the ones the client generates...so if you don't know that and have only one queue (*map*, actually!) you are going to confuse, collide, overwrite and lose packets of both types, instantly breaking QoS handshaking. The other lib only works on the "Golden path" where nothing breaks or needs re-transmission. It clears the session unconditionally on connect , instantly breaking any chance of allowing outbound Qos retries....

But hell, his code sure *looks* prettier than mine.

Rest assured, gentlemen, I would not write code like this professionally if I was in or managing a team! In fact if anyone *did* do so, I'd probably have the same constructive and partially valid criticisms as you have, so, sincerely, "I hear you". But *this* team has -a nd only ever will have - one member: Me. So it's "*my* way or the highway", I'm afraid!

I am grateful for your support and hope to keep you here instead of "over there" so don't take this too seriously or even "the wrong way" but I have been writing code for 45 years, so even if I know *some* of what you say is valid , I'm hardly likely to change now, am I? I'm also of the view that code like this solo project is only ever going to be maintained by me, so as long as *I* can read it I really don't care two hoots if no-one else can of if they think "aaaaaaargh" that' looks horrible. It is not written for its looks, its written for *ME* to be able to get it out quickly and above all, *working*.

The rest of my theory is that if anyone IS smart enough to know and make those (sometimes) valid comments (you guys obvioulsy are! :) ) then they must understand how the code works, so a) whats the problem? b) they already know enough it fix it themselves - its open source: you dont like, fork it! :) :) :) Them's the rules.

Please take all of this with big smile, I'm happy to take criticism re portability, potential unseen problems, genuine bugs and *functionality* etc, but style? No. Never. I used to *teach* it and enforce it, so I now choose the stye I like and works fast for me - it ain't a beauty contest...the ONLY valid response back to both of you is:

"YES, BUT DOES IT WORK?"

One day in the future when its rock-solid stable - with your help - and everyone in the world uses it, we can share a beer and argue over what color lipstick to put on this pig: it will still always be a pig. Till then, it's staying as ugly and functional as it is. I have three other cross-platform projects (STM32) that are going to rely on it and are HUGELY delayed by my very bad decision to assume that other lib actually worked. So I'm running late, I'm 65, grumpy, and set in my ways, so let's not waste time on "style" that i used and discarded 30 years ago, eh?

You want to see truly bad code? FFS look at the other lib: one of the worst, most shocking, inefficent erro-prone bug-ridden piles of crap I have ever seen. The bloke doesn't have A CLUE about C. I bet he learned java or c# at college, amd this was a project that got badly out of hand.

Anyway, I'm happy being a grumpy old man. I love your comments, but I'm too busy right now to do anything unless its impacts *functionality*. I'm not saying I won't be in the future, so keep on coming, and thanks again.

PS: I'm not ALWAYS this grumpy!!!!!!!!!!!!!!!!!!! I'm just seriously annoyed at the amount of time and effort that other bullshit library caused me. The truth is I'm angry at myself for not checking it out and/or evaluating it more thoroughly.

So while these comments are somewhat valid, I don't see any of them as "issues" hence I will close this. The best place for subsequent discussions like this are my facebook groups which I monitor constantly: 
https://www.facebook.com/groups/2125820374390340/?ref=bookmarks
https://www.facebook.com/groups/h4plugins/
https://www.facebook.com/groups/esp8266questions/
or https://www.facebook.com/groups/ESP8266/ which I moderate

Please come and join any/all that might be relevant to you, and *spread the word*: the more pople on this project the quicker we will gets the bugs out and have a *useful* and stable library. Even if it's an ugly one :)

Phil

**Comment by bertmelis** [comment]
1) Calm down, it's going in the right direction. You should be happy to see light at the end of the tunnel.
For the record: the author of the other lib was -at the time he wrote the lib- completely self-taught. So I'm only blaming myself not to see the obvious.

2) Style is indeed just style. As long as it's consistent it's fine by me.

3) So if I understand correctly, single TCP-packet messages are currently supported, for all QoS levels?

**Comment by proddy** [comment]
Agree, and I will also support and contribute to this eventual replacement of Marvin's original async mqtt library. To give the guy some credit he did write it 4 years ago when he was a kid learning C++. It's only a shame that it hasn't really evolved and the numerous feedbacks and PRs largely ignored.

My personal viewpoint: I do think style and clean code is important, especially when others need to understand and contribute. Readability over lines of code. The object size and eventual RAM is a good measurement as these are sparse on MCUs. And using C++ containers like std::string is IMO a good thing. Arduino's Strings library is evil.

**Comment by obrain17** [comment]
I will also support and contribute, as I already announced in [https://github.com/marvinroger/async-mqtt-client/issues/193](https://github.com/marvinroger/async-mqtt-client/issues/193).
Unfortunately I did not see this issue-thread #1, so I created the PR #2  with some suggestions.

I will replace marvin's library with yours on one ESP8266 and one ESP32 device. I am using platformio IDE, that is why I suggested some changes to get rid of warnings which would obscure important ones.

Unfortunately I am not a facebook member and would prefer a discussion here in github.
